server-common/src/main/java/com/datastrato/gravitino/server/auth/KerberosAuthenticator.java

```diff
@@ -0,0 +1,183 @@
+
+  public static final Logger LOG = LoggerFactory.getLogger(KerberosAuthenticator.class);
+  private final Subject serverSubject = new Subject();
+  private GSSManager gssManager;
+
+  @Override
+  public void initialize(Config config) throws RuntimeException {
+    try {
+      String principal = config.get(KerberosConfig.PRINCIPAL);
+      String keytab = config.get(KerberosConfig.KEYTAB);
+      File keytabFile = new File(keytab);
+      if (!keytabFile.exists()) {
+        throw new IllegalArgumentException(String.format("Keytab %s doesn't exist", keytab));
+      }
+
+      if (!keytabFile.canRead()) {
+        throw new IllegalArgumentException(String.format("Keytab %s can't be read", keytab));
+      }
+
+      Principal krbPrincipal = new KerberosPrincipal(principal);
+      LOG.info("Using keytab {}, for principal {}", keytab, krbPrincipal);
+      serverSubject.getPrincipals().add(krbPrincipal);
+      KeyTab keytabInstance = KeyTab.getInstance(keytabFile);
+      serverSubject.getPrivateCredentials().add(keytabInstance);
+
+      gssManager =
+          Subject.doAs(
+              serverSubject,
+              new PrivilegedExceptionAction<GSSManager>() {
+                @Override
+                public GSSManager run() throws Exception {
+                  return GSSManager.getInstance();
+                }
+              });
+    } catch (PrivilegedActionException ex) {
+      throw new RuntimeException(ex);
+    }
+  }
+
+  @Override
+  public boolean isDataFromToken() {
+    return true;
+  }
+
+  @Override
+  public Principal authenticateToken(byte[] tokenData) {
+    if (tokenData == null) {
+      throw new UnauthorizedException("Empty token authorization header", AuthConstants.NEGOTIATE);
+    }
+
+    String authData = new String(tokenData);
+    if (StringUtils.isBlank(authData)
+        || !authData.startsWith(AuthConstants.AUTHORIZATION_NEGOTIATE_HEADER)) {
+      throw new UnauthorizedException(
+          "Invalid token authorization header", AuthConstants.NEGOTIATE);
+    }
+
+    String token = authData.substring(AuthConstants.AUTHORIZATION_NEGOTIATE_HEADER.length());
+    byte[] clientToken = Base64.getDecoder().decode(token);
+    if (StringUtils.isBlank(token)) {
+      throw new UnauthorizedException("Blank token found", AuthConstants.NEGOTIATE);
+    }
+
+    try {
+      String serverPrincipal = KerberosServerUtils.getTokenServerName(clientToken);
+      return Subject.doAs(
+          serverSubject,
+          new PrivilegedExceptionAction<Principal>() {
+            @Override
+            public Principal run() throws Exception {
+              return validateClientToken(serverPrincipal, clientToken);
+            }
+          });
+    } catch (Exception e) {
+      throw new UnauthorizedException("Fail to validate the token", AuthConstants.NEGOTIATE);
+    }
+  }
+
+  private Principal validateClientToken(String serverPrincipal, byte[] clientToken)
```

Comment:
This method doesn't only validate the token, but also return the user principal if the check is passed. I would suggest use a better method name.

Reply:
How about `retrievePrincipalFromToken`?